CenterForOpenScience/waterbutler
# [SVCS-547] Ft/s3 chunked uploads

## Ticket

https://openscience.atlassian.net/browse/SVCS-547

The rest of this description is copied from https://github.com/CenterForOpenScience/waterbutler/pull/294 - John's original PR. This PR has been updated to ensure removal of partial incomplete uploads, and also tweaks the chunk uploading process to upload in parallel

## Purpose

This allows very large files to be uploaded via chunks using the S3 addon.

## Changes

Adds unit tests and three major methods to the S3 provider which (1) creates an upload session (2) uploads parts to the session and (3) "commits" the parts and completes the session.

## Side effects

None that I know of.

## QA Notes

Uploading large files via Fangorn will require a small osf-side fix - there's frontend validation on the file size. https://github.com/CenterForOpenScience/osf.io/pull/8406
Need to make sure uploaded files are downloaded and are equivalent (Make sure parts are in the right order)


## Deployment Notes

No special deployment notes.
